### ISSUE TITLE
`(x for x in xs if await ..)` not detected as an async generator

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#3723]
Generator expressions that ["contain `await` expressions"](https://docs.python.org/3/reference/expressions.html#generator-expressions) are async generators, but ty doesn't recognize them as such:

```py
import asyncio
from collections.abc import AsyncIterator


async def is_even(value: int) -> bool:
    return value % 2 == 0


async def main() -> None:
    generator = (value for value in [1, 2, 3] if await is_even(value))

    await anext(generator, None)  # ty: Argument to function `anext` is incorrect: Expected `SupportsAnext[Unknown]`, found `GeneratorType[int, None, None]`
```

**Comment by AlexWaygood** [comment]
here's some analysis by codex:

<details>

# When does `await` make a generator expression asynchronous?

The [generator-expression documentation](https://docs.python.org/3/reference/expressions.html#generator-expressions) states that a generator expression containing an `async for` clause or an `await` expression is an asynchronous generator expression. The important question is what "contains" means, particularly when the `await` is nested inside another expression or appears in a nested comprehension.

## Conclusion

"Contains" is best understood as "is evaluated in the generator expression's implicit scope", not simply "appears textually between the generator expression's parentheses".

An `await` can be nested arbitrarily deeply inside ordinary expressions evaluated in the generator expression's scope. For example, all of these produce asynchronous generators:

```python
(await f(x) for x in xs)
([await f(x)] for x in xs)
(x for x in xs if await predicate(x))
(y for x in xs for y in await f(x))
```

There are two important scope-related qualifications:

1. The iterable expression in the leftmost `for` clause is evaluated in the enclosing scope, so an `await` anywhere in that expression does not make the generator expression asynchronous.
2. A nested generator expression has its own generator scope. If the nested generator expression is asynchronous, that does not make the outer generator expression asynchronous. By contrast, a nested asynchronous list, set, or dictionary comprehension does make the outer comprehension asynchronous because evaluating the nested container comprehension itself requires suspension.

Consequently:

| Location of asynchronous operation | Result for the outer generator expression |
|---|---|
| `await` in the yielded element, including inside ordinary nested expressions | Asynchronous |
| `await` in an `if` filter | Asynchronous |
| `await` in the iterable of a later `for` clause | Asynchronous |
| `async for` in any clause | Asynchronous |
| `await` in the iterable of the leftmost `for` clause | Synchronous, unless the clause itself is `async for` |
| Nested asynchronous list, set, or dictionary comprehension | Asynchronous, propagated outward |
| Nested asynchronous generator expression | Synchronous outer generator yielding asynchronous inner generators |

For example:

```python
async def make_generators(xs, ys):
    sync_gen = (x for x in await get_xs())
    async_gen = ([await f(x)] for x in xs)
    sync_outer = ((await f(y) for y in ys) for x in xs)
    return sync_gen, async_gen, sync_outer
```

Here, `sync_gen` is synchronous because `await get_xs()` is evaluated before the generator expression's implicit scope is entered. `async_gen` is asynchronous because its `await` is evaluated in that implicit scope. `sync_outer` is synchronous because constructing each inner asynchronous generator does not itself suspend the outer generator.

## Documentation

The generator-expression-specific paragraph uses the broad wording "contains either `async for` clauses or `await` expressions" in [`Doc/reference/expressions.rst`, lines 987-991](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Doc/reference/expressions.rst#L987-L991).

The earlier general section on asynchronous comprehensions states the boundary more precisely: an asynchronous operation counts anywhere except the iterable expression in the leftmost `for` clause. See [`Doc/reference/expressions.rst`, lines 905-916](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Doc/reference/expressions.rst#L905-L916). It also records that, since Python 3.11, asynchronous comprehensions can be nested and outer comprehensions implicitly become asynchronous; see [`Doc/reference/expressions.rst`, lines 922-925](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Doc/reference/expressions.rst#L922-L925).

## Direct generator-expression tests

### `await` in a filter

`test_async_gen_expression_02` is the clearest direct test of an `await` changing a generator expression into an asynchronous generator. It defines a synchronous function returning this expression:

```python
(i * 2 for i in range(n) if await wrap(i))
```

The expression is then consumed with `async for`. See [`Lib/test/test_asyncgen.py`, lines 1823-1836](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Lib/test/test_asyncgen.py#L1823-L1836). This proves that an `await` in a filter makes the generator expression asynchronous. It also demonstrates that, since Python 3.7, such an expression may be created inside a synchronous function because the `await` belongs to the generator expression's implicit scope rather than to the enclosing function.

### `await` in the yielded element

`test_builtin_call_async_genexpr_no_crash` passes expressions such as the following to synchronous builtins:

```python
all(await 2 for _ in [])
```

The builtins reject the resulting asynchronous generator as a synchronous iterable, and the test checks for `TypeError`. See [`Lib/test/test_builtin.py`, lines 311-330](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Lib/test/test_builtin.py#L311-L330). This supplies direct coverage for an `await` in the yielded element.

### `async for`

`test_async_gen_expression_01` verifies the unambiguous `async for` case:

```python
(i * 2 async for i in arange(n))
```

See [`Lib/test/test_asyncgen.py`, lines 1807-1821](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Lib/test/test_asyncgen.py#L1807-L1821).

## Nested-comprehension tests

`test_nested_comp` provides the most useful evidence about what happens across nested implicit scopes.

### Asynchronous container comprehension inside a generator expression

`run_list_inside_gen` constructs a generator expression whose yielded element is an asynchronous list comprehension:

```python
gen = ([i + j async for i in asynciter([1, 2])] for j in [10, 20])
return [x async for x in gen]
```

The outer generator expression is consumed with `async for`, demonstrating that the nested asynchronous list comprehension makes the outer generator expression asynchronous. See [`Lib/test/test_coroutines.py`, lines 2081-2086](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Lib/test/test_coroutines.py#L2081-L2086).

### Asynchronous generator expression inside a generator expression

`run_gen_inside_gen` constructs a synchronous outer generator expression whose yielded elements are asynchronous generator expressions:

```python
gens = ((i async for i in asynciter(range(j))) for j in [3, 5])
return [x for g in gens async for x in g]
```

The ordinary `for g in gens` demonstrates that `gens` itself is synchronous. Each `g` is an asynchronous generator, but that asynchronous nature does not propagate to the outer generator expression. See [`Lib/test/test_coroutines.py`, lines 2095-2100](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Lib/test/test_coroutines.py#L2095-L2100).

The contrast between these two cases is intentional: evaluating an asynchronous container comprehension requires waiting for the completed container, whereas evaluating an asynchronous generator expression only creates an asynchronous generator object.

## Supporting comprehension tests

The implementation path is shared by generator expressions and list, set, and dictionary comprehensions. `test_comp_1` covers `await` in the produced element of container comprehensions; see [`Lib/test/test_coroutines.py`, lines 1847-1866](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Lib/test/test_coroutines.py#L1847-L1866).

`test_comp_2` covers `await` in later iterable expressions, including multiple nested `for` clauses; see [`Lib/test/test_coroutines.py`, lines 1868-1898](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Lib/test/test_coroutines.py#L1868-L1898). These are not generator-expression tests themselves, but they exercise the same symbol-table traversal responsible for deciding whether a comprehension is asynchronous.

## Compiler implementation

The compiler's symbol-table pass supplies the precise definition of "contains".

### The leftmost iterable is outside the implicit scope

`symtable_handle_comprehension()` first visits the entire iterable expression of the leftmost clause in the current, enclosing scope. Only after doing so does it enter the implicit comprehension scope. See [`Python/symtable.c`, lines 3082-3097](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Python/symtable.c#L3082-L3097).

The iteration target, filters belonging to the first clause, all later clauses, the dictionary value when applicable, and the yielded element are then visited inside the implicit scope. See [`Python/symtable.c`, lines 3121-3131](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Python/symtable.c#L3121-L3131).

This explains why the whole subtree of the leftmost iterable is excluded, rather than only an `await` immediately following `in`.

### Ordinary expression nesting is recursive

The symbol-table visitor recursively traverses ordinary expression nodes. For example, it visits every operand of boolean, binary, unary, and conditional expressions; see [`Python/symtable.c`, lines 2485-2524](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Python/symtable.c#L2485-L2524). When it reaches an `Await` node in the current scope, it sets `ste_coroutine`; see [`Python/symtable.c`, lines 2570-2590](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Python/symtable.c#L2570-L2590).

The decision is therefore syntactic and scope-based. An `await` in an ordinary nested expression marks the generator expression as asynchronous even if control flow means that the `await` will not execute on a particular iteration.

### `async for` and nested propagation

When the visitor encounters an `async for` clause, it marks the current comprehension scope as a coroutine; see [`Python/symtable.c`, lines 3057-3069](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Python/symtable.c#L3057-L3069). The leftmost clause is handled separately but has the same effect when it is `async for`; see [`Python/symtable.c`, lines 3112-3114](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Python/symtable.c#L3112-L3114).

After visiting a nested comprehension, coroutine status is propagated to the enclosing comprehension only when the nested comprehension is not a generator expression. The `!is_generator` condition is what creates the distinction between nested container comprehensions and nested generator expressions. See [`Python/symtable.c`, lines 3131-3148](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Python/symtable.c#L3131-L3148).

Finally, a code object marked as both a generator and a coroutine receives the `CO_ASYNC_GENERATOR` flag; see [`Python/compile.c`, lines 1424-1448](https://github.com/python/cpython/blob/f2a0f82282d6307f7fd2d4ccf52a8fd95ac3922f/Python/compile.c#L1424-L1448).

## Test coverage gaps

The existing test suite establishes the central behavior, but it does not provide a comprehensive location-by-location matrix.

In particular, there does not appear to be a checked-in test that directly asserts that this expression remains synchronous:

```python
(x for x in await get_xs())
```

There also does not appear to be a parameterized generator-expression test covering arbitrary ordinary nesting such as calls, conditional expressions, container displays, and subscripts. Those conclusions follow from the documented leftmost-iterable exception and the recursive symbol-table implementation rather than from dedicated regression tests for every expression shape.

</details>

It's conclusion is:

> "Contains" is best understood as "is evaluated in the generator expression's implicit scope", not simply "appears textually between the generator expression's parentheses".

So if `await` appears in the expression being iterated over, that expression is evaluated in the context of an outer scope, so it does not make the generator expression asynchronous, e.g. `(x for x in await foo)` is not an async generator. And if `await` appears in a nested generator/comprehension expression, that also does not make the generator async, because those nested expressions create nested scopes: `((await x for x in y) for y in range(42))` is also not an async generator. But _here_, the outer generator _is_ async, because the expression being iterated over by the inner generator contains an `await`, and that expression is evaluated in the scope of the outer generator:

```pycon
>>> def f():
...     x = ((x for x in await y) for y in range(42))
...     print(type(x))
...
>>> f()
<class 'async_generator'>
```

**Comment by mvanhorn** [comment]
Implemented in https://github.com/astral-sh/ty/pull/3795 - fix: Detect generator expressions containing await as async generators.
